dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +241,260 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
-
-                winrt::hstring taskClsidStr{ winrt::to_hstring(taskClsid) };
-                winrt::hstring backgroundTaskFullName{ backgroundTaskName + taskClsidStr };
-
-                auto tasks{ BackgroundTaskRegistration::AllTasks() };
-                bool isTaskRegistered{ std::any_of(std::begin(tasks), std::end(tasks),
-                    [&](auto&& task)
+                    winrt::hstring backgroundTaskFullName{ backgroundTaskName + winrt::to_hstring(m_registeredClsid) };
+                    if (!IsBackgroundTaskRegistered(backgroundTaskFullName))
                     {
-                        auto name{ task.Value().Name() };
+                        BackgroundTaskBuilder builder{ BackgroundTaskBuilder() };
+
+                        builder.Name(backgroundTaskFullName);
 
-                        if (std::wstring_view(name).substr(0, backgroundTaskName.size()) != backgroundTaskName)
-                        {
-                            return false;
-                        }
+                        PushNotificationTrigger trigger{};
+                        builder.SetTrigger(trigger);
 
-                        if (name == backgroundTaskFullName)
-                        {
-                            s_pushTriggerRegistration = task.Value();
-                            return true;
-                        }
+                        auto builder5{ builder.as<winrt::IBackgroundTaskBuilder5>() };
+                        builder5.SetTaskEntryPointClsid(m_registeredClsid);
+ 
+                        m_pushTriggerRegistration = builder.Register();
+                    }
+                }
 
-                        throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");
-                    }) };
+                // Register a sink for the application to receive foreground raw notifications
+                RegisterForegroundSinkHelper();
 
-                if (!isTaskRegistered)
+                // Register a PushNotificationBackgroundTask to handle background activation scenarios
+                if (!m_comActivatorRegistration)
                 {
-                    builder = BackgroundTaskBuilder();
-                    builder.Name(backgroundTaskFullName);
-
-                    PushNotificationTrigger trigger{};
-                    builder.SetTrigger(trigger);
+                    THROW_IF_FAILED(::CoRegisterClassObject(
+                        m_registeredClsid,
+                        winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                        CLSCTX_LOCAL_SERVER,
+                        REGCLS_MULTIPLEUSE,
+                        &m_comActivatorRegistration));
+                }
 
-                    THROW_HR_IF(E_NOTIMPL, !AppModel::Identity::IsPackagedProcess());
+                scopeExitToCleanRegistrations.release();
+            }
+            else
+            {
+                auto scopeExitToCleanRegistrations{ wil::scope_exit([&]()
```

Comment:
You need to throw here if Register is already done. We wont allow 2 registrations